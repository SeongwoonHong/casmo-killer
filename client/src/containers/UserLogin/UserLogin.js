import React, { Component } from 'react';

import * as storage from '@sharedUtils/storage';
import FormMessage from '@sharedComponents/FormMessage';

import SocialLogin from '../../components/SocialLogin';
import LocalLogin from '../../components/LocalLogin';

import './UserLogin.scss';

class Login extends Component {

  componentWillReceiveProps(nextProps) {

    const { history, user } = nextProps;

    if (user.isLoggedIn) {

      history.replace(this.props.location && this.props.location.state
        ? this.props.location.state.from
        : '/'
      );

    }

  }

  onSocialRegister = (userInfo) => {

    this.props.setUserForRegistration(userInfo);

    const { auth, history } = this.props;

    if (auth.email) {
      history.push('/user/register');
    }

  };

  onLoginSuccess = async (userInfo) => {

    try {

      const user = await storage.set('ckUser', userInfo);
      this.props.loginSuccess(user, true);

    } catch (error) {

      console.error(error);

    }

  };

  render() {

    const { location, match, auth } = this.props;

    const isLogin = match.params.type === 'login';
    const redirectUrl = location && location.state ? location.state.from : '/';

    return (
      <div className="Login">
        <h2 className="User__page__title">
          {
            isLogin
            ? 'Login to your account'
            : 'Sign up for free'
          }
          <i className="User__page__title__icons material-icons">
            {
              isLogin
              ? 'lock'
              : 'mode_edit'
            }
          </i>
        </h2>
        <FormMessage
          className="Login__message"
          message={ auth.message } />
        <SocialLogin
          isLogin={ isLogin }
          onRegister={ this.onSocialRegister }
          onSuccess={ this.onLoginSuccess } />
        <LocalLogin
          isLogin={ isLogin }
          redirectUrl={ redirectUrl }
          onSuccess={ this.onLoginSuccess } />
      </div>
    );
  }

}

export default Login;
