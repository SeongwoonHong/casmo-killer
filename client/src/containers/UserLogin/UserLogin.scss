.Login {
    > .Form-message {
        margin: 0 15px 15px !important;
    }
}
