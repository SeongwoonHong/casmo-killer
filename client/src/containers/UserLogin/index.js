import { connect } from 'react-redux';
import * as actions from '@actions';

import UserLogin from './UserLogin';

function mapStateToProps(state) {
  return {
    auth: state.auth,
    user: state.user.user
  };
}

function mapDispatchToProps(dispatch) {
  return {
    loginSuccess: (payload, shouldReset) => dispatch(actions.loginSuccess(payload, shouldReset)),
    setUserForRegistration: userInfo => dispatch(actions.setUserForRegistration(userInfo))
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(UserLogin);
