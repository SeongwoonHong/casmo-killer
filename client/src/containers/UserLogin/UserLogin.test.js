/* eslint react/jsx-boolean-value: 0 */
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { mount, shallow } from 'enzyme';
import sinon from 'sinon';
import configureStore from 'redux-mock-store';

import UserLogin from './UserLogin';
import ConnectedUserLogin from './index';

const mockStore = configureStore();

describe('<UserLogin />', () => {

  let component;

  beforeEach(() => {

    const auth = { message: '' };
    const user = { isLoggedIn: false };

    component = shallow(
      <UserLogin
        auth={ auth }
        user={ user }
        match={ { params: { type: 'login' } } } />
    );

  });

  it('should render the component consistently', () => {
    expect(component).toMatchSnapshot();
  });

  it('should display correct text and icons', () => {
    expect(component.find('.user-page-title').text()).toEqual('Login to your accountlock');
    component.setProps({ match: { params: { type: 'register' } } });
    expect(component.find('.user-page-title').text()).toEqual('Sign up for freemode_edit');
  });

  it('should trigger componentWillReceiveProps when login status changes', () => {

    const auth = { message: '' };
    const user = { isLoggedIn: false };

    const spy = sinon.spy(UserLogin.prototype, 'componentWillReceiveProps');
    const replace = sinon.spy();

    const wrapper = shallow(
      <UserLogin
        auth={ auth }
        user={ user }
        history={ { replace } }
        match={ { params: { type: 'login' } } } />
    );

    wrapper.setProps({ user: { isLoggedIn: true } });

    expect(spy.calledOnce).toEqual(true);
    expect(replace.calledWith('/')).toEqual(true);

  });

  it('should trigger loginSuccess action when login is successful', async () => {

    const loginSuccess = sinon.spy();

    const userData = { user: { isLoggedIn: false } };
    const promise = Promise.resolve(userData);

    const wrapper = shallow(
      <UserLogin
        auth={ { message: '' } }
        user={ userData }
        loginSuccess={ loginSuccess }
        match={ { params: { type: 'login' } } } />
    );

    wrapper.instance().onLoginSuccess(userData);

    await promise;

    expect(loginSuccess.calledWith(userData, true)).toEqual(true);

  });

  it('should trigger setUserForRegistration action when it is social register', () => {

    const setUserForRegistration = sinon.spy();

    const userData = { user: { strategy: 'facebook' } };

    const wrapper = shallow(
      <UserLogin
        auth={ { message: '' } }
        user={ userData }
        setUserForRegistration={ setUserForRegistration }
        match={ { params: { type: 'login' } } } />
    );

    wrapper.instance().onSocialRegister(userData);
    expect(setUserForRegistration.calledWith(userData)).toEqual(true);

  });

});

describe('<ConnectedUserLogin />', () => {

  const initialState = {
    auth: {
      message: ''
    },
    user: {
      user: {
        isLoggedIn: false
      }
    }
  };
  const store = mockStore(initialState);

  let component;

  beforeEach(() => {

    component = shallow(
      <ConnectedUserLogin store={ store } />
    );

  });

  it('should render the component consistently', () => {
    expect(component).toMatchSnapshot();
  });

  it('should receive correct props from the store', () => {
    expect(component.prop('auth')).toEqual(initialState.auth);
    expect(component.prop('user')).toEqual(initialState.user.user);
  });

});
