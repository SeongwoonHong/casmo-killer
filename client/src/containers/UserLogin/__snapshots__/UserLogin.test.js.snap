// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConnectedUserLogin /> should render the component consistently 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Login)
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "auth": Object {
        "message": "",
      },
      "loginSuccess": [Function],
      "setUserForRegistration": [Function],
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
      "user": Object {
        "isLoggedIn": false,
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "auth": Object {
          "message": "",
        },
        "loginSuccess": [Function],
        "setUserForRegistration": [Function],
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
        "user": Object {
          "isLoggedIn": false,
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<UserLogin /> should render the component consistently 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    auth={
        Object {
            "message": "",
          }
    }
    match={
        Object {
            "params": Object {
              "type": "login",
            },
          }
    }
    user={
        Object {
            "isLoggedIn": false,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2
          className="User__page__title"
>
          Login to your account
          <i
                    className="User__page__title__icons material-icons"
          >
                    lock
          </i>
</h2>,
        <FormMessage
          className="Login__message"
          message=""
          type="error"
/>,
        <SocialLogin
          isLogin={true}
          onRegister={[Function]}
          onSuccess={[Function]}
/>,
        <LocalLogin
          isLogin={true}
          onSuccess={[Function]}
          redirectUrl="/"
/>,
      ],
      "className": "Login",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Login to your account",
            <i
              className="User__page__title__icons material-icons"
>
              lock
</i>,
          ],
          "className": "User__page__title",
        },
        "ref": null,
        "rendered": Array [
          "Login to your account",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "lock",
              "className": "User__page__title__icons material-icons",
            },
            "ref": null,
            "rendered": "lock",
            "type": "i",
          },
        ],
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "className": "Login__message",
          "message": "",
          "type": "error",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "isLogin": true,
          "onRegister": [Function],
          "onSuccess": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "isLogin": true,
          "onSuccess": [Function],
          "redirectUrl": "/",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2
            className="User__page__title"
>
            Login to your account
            <i
                        className="User__page__title__icons material-icons"
            >
                        lock
            </i>
</h2>,
          <FormMessage
            className="Login__message"
            message=""
            type="error"
/>,
          <SocialLogin
            isLogin={true}
            onRegister={[Function]}
            onSuccess={[Function]}
/>,
          <LocalLogin
            isLogin={true}
            onSuccess={[Function]}
            redirectUrl="/"
/>,
        ],
        "className": "Login",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Login to your account",
              <i
                className="User__page__title__icons material-icons"
>
                lock
</i>,
            ],
            "className": "User__page__title",
          },
          "ref": null,
          "rendered": Array [
            "Login to your account",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "lock",
                "className": "User__page__title__icons material-icons",
              },
              "ref": null,
              "rendered": "lock",
              "type": "i",
            },
          ],
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "Login__message",
            "message": "",
            "type": "error",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "isLogin": true,
            "onRegister": [Function],
            "onSuccess": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "isLogin": true,
            "onSuccess": [Function],
            "redirectUrl": "/",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
